File: python_port/src/nls_python/model_generation/inference_rules/eq_elim.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Set, Tuple

from nls_python.typing_utils import typechecked

from nls_python.fol.formula.equals import Equals
from nls_python.fol.formula import Formula
from nls_python.fol.substitution import Substitution
from nls_python.fol.unification import unify

from ..tableau_model import TableauModel
from .base import InferenceRule, RuleApplication


@typechecked
class EqElim(InferenceRule):
    @property
    def branching(self) -> bool:
        return False

    def apply(self, model: TableauModel) -> bool:
        applications = self._collect_applications(model)
        if not applications:
            return False
        for application in applications:
            model.apply_substitution(application.substitution)
            model.add_rule_applications({application})
        return True

    def _collect_applications(self, model: TableauModel) -> Set["EqElimApplication"]:
        applications: Set[EqElimApplication] = set()
        seen_products: Set[Formula] = set()
        current_sub = model.get_substitution()
        for formula in model.get_formulas():
            if not isinstance(formula, Equals):
                continue
            maybe_sub = unify(formula.left, formula.right, model.get_substitution())
            if maybe_sub is None:
                continue
            composed = current_sub.compose(maybe_sub)
            outputs = {
                f.apply_sub(composed)
                for f in model.get_formulas()
                if f.apply_sub(composed) not in model.get_formulas()
            }
            outputs -= seen_products
            seen_products.update(outputs)
            application = EqElimApplication(formula, maybe_sub, tuple(outputs))
            if application in model.get_rule_applications():
                continue
            current_sub = composed
            applications.add(application)
        return applications


@typechecked
@dataclass(frozen=True)
class EqElimApplication(RuleApplication):
    source: Equals
    substitution: Substitution
    outputs: Tuple[Formula, ...] = field(default_factory=tuple)

    def __hash__(self) -> int:
        return hash(self.source)

    def __eq__(self, other: object) -> bool:
        return isinstance(other, EqElimApplication) and self.source == other.source

    def __str__(self) -> str:
        outputs = ", ".join(str(formula) for formula in self.outputs) or "∅"
        return f"EqElim: {self.source} using {self.substitution} -> [{outputs}]"

File: impl_java/modelGeneration/folModel/rules/EqElim.java
package modelGeneration.folModel.rules;

import fol.Substitution;
import fol.Unifier;
import fol.formula.Equals;
import fol.formula.Formula;
import modelGeneration.folModel.Model;

import java.util.HashSet;
import java.util.Set;
import java.util.stream.Collectors;

public class EqElim implements InferenceRule {
    @Override
    public boolean apply(final Model model) {
        var applications = getRuleApplications(getEqs(model), model);
        if (applications.isEmpty()) return false;
        for(var application : applications){
            model.applySubstitution(application.sub());
            model.addRuleApplications(Set.of(application));
        }
        return true;
    }

    public Set<Equals> getEqs(Model model) {
        return model.getFormulas().stream()
                .filter(Equals.class::isInstance)
                .map(Equals.class::cast)
                .collect(Collectors.toSet());
    }

    public Set<EqElimRuleApplication> getRuleApplications(Set<Equals> eqForms, Model model) {
        Set<Formula> allProducts = new HashSet<>();
        Set<EqElimRuleApplication> applications = new HashSet<>();
        Substitution sub = model.getSubstitution();
        var modelFormulas = model.getFormulas();
        for (Equals eq : eqForms) {
            var maybeSub = Unifier.unify(eq.left(), eq.right(), model.getSubstitution());
            if(maybeSub.isEmpty())
                continue;
            final var innerSub = sub.compose(maybeSub.get());
            var emptyApplication = new EqElimRuleApplication(eq, maybeSub.get(), Set.of());
            if(model.getRuleApplications().contains(emptyApplication)){
                continue;
            }
            var out = modelFormulas.stream()
                    .map(f -> f.applySub(innerSub))
                    .filter(f -> !allProducts.contains(f) && !modelFormulas.contains(f))
                    .collect(Collectors.toSet());
            allProducts.addAll(out);
            applications.add(new EqElimRuleApplication(eq, maybeSub.get(), out));
            sub = innerSub;
        }
        return applications;
    }

    @Override
    public boolean isBranching() {
        return false;
    }

    public record EqElimRuleApplication(Equals input, Substitution sub, Set<Formula> output) implements RuleApplication {

        @Override
        public String toString() { return getString(0, ""); }

        @Override
        public boolean equals(Object obj) {
            if (obj instanceof EqElimRuleApplication(Equals input1, Substitution _, Set<Formula> _)) {
                return input.equals(input1);
            }
            return false;
        }

        @Override
        public String getString(int indentation, String delim) {
            if (indentation == 0)
                return "EqElim : " + input + " -> " + output;
            String indentString = "  ".repeat(indentation) + delim;
            return indentString + " EqElim: " + input + " -> " + output;
        }

        @Override
        public int hashCode() {
            return input.hashCode();
        }
    }
}
